OpenTitan: Update the HMAC Implementation

alistair23 opened the pull request:
### Pull Request Overview

This PR does a few different things.

First I convert the key length to be unspecified. This means we can allow variable length keys.

Secondly I convert the HMAC digest to use const generics.

Then I have stared testing the HMAC capsule on OT after the 2.0 upgrade. This fixes some issues there, including double entering grants.

Finally some hardware related fixes.

### Testing Strategy

Run libtock-c hmac operation on OpenTitan FPGA.

### TODO or Help Wanted

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

hudson-ayers commented on chips/lowrisc/src/hmac.rs:
```diff
@@ -269,3 +277,15 @@ impl hil::digest::HMACSha256 for Hmac<'_> {
     }
 }
+
+impl hil::digest::HMACSha384<32> for Hmac<'_> {
+    fn set_mode_hmacsha384(&self, _key: &[u8; 32]) -> Result<(), ErrorCode> {
+        Err(ErrorCode::NOSUPPORT)
+    }
+}
+
+impl hil::digest::HMACSha512<32> for Hmac<'_> {
+    fn set_mode_hmacsha512(&self, _key: &[u8; 32]) -> Result<(), ErrorCode> {
+        Err(ErrorCode::NOSUPPORT)
+    }
+}
```
I don't love the pattern of implementing traits to just return errors, but I understand the design of the virtualizer sort of requires this (at least without adding a decent amount of complexity).

hudson-ayers reviewed (commented):
Looks good for the most part, always nice to see more use of const generics. I have a few comments

alistair23 commented on chips/earlgrey/src/chip.rs:
```diff
@@ -76,6 +76,10 @@ impl<'a> InterruptService<()> for EarlGreyDefaultPeripherals<'a> {
                 self.i2c0.handle_interrupt()
             }
+            interrupts::KMAC_KMACDONE => {
+                // Either a hardware or documentation bug
```
Yeah, we only ever use the HMAC in Tock, but it is generating KMAC interrupts

alistair23 commented on chips/earlgrey/src/chip.rs:
```diff
@@ -110,6 +114,11 @@ impl<'a, A: 'static + Alarm<'static>, I: InterruptService<()> + 'a> EarlGrey<'a,
 
     unsafe fn handle_plic_interrupts(&self) {
         while let Some(interrupt) = self.plic.get_saved_interrupts() {
+            self.atomic(|| {
+                // Safe as interrupts are disabled
+                self.plic.complete(interrupt);
+            });
+
```
Yep, I have added this in the code as well:

```rust
            // Claim the interrupt before we handle it.
            // Currently the interrupt has been claimed but not completed.
            // This means that if the interrupt re-asserts we will loose the
            // re-assertion. Generally this isn't a problem, but some of the
            // interrupt handlers expect that interrupts could occur.
            // For example the HMAC interrupt handler will write data to the
            // HMAC buffer. We then rely on an interrupt triggering when that
            // buffer comes empty. This can happen while we are still in the
            // interrupt handler. To ensure we don't loose the interrupt we
            // claim it here.
            // In order to stop an interrupt loop, we first disable the
            // interrupt.
```

alistair23 commented:
Thanks for the review. I addressed your comments. I also changes the structure a little bit. I realised using const generics for the key length meant that we only support a single key length. Although that's all OT supports I suspect there will be hardware that can handle a range of lengths. To handle that I made the key length dynamic.

hudson-ayers reviewed (dismissed):
It might be nice for someone else to take a look at the change to interrupt handling, as I know there can be some very subtle bugs there, but this looks good to me.

phil-levis commented on chips/earlgrey/src/chip.rs:
```diff
@@ -110,6 +114,25 @@ impl<'a, A: 'static + Alarm<'static>, I: InterruptService<()> + 'a> EarlGrey<'a,
+            // In order to stop an interrupt loop, we first disable the
+            // interrupt. `service_pending_interrupts()` will re-enable
+            // interrupts once they are all handled.
+            self.atomic(|| {
```
I am a little worried about this logic.

In many cases, it seems like it's OK for the interrupt to be completed after the full handler. In the HMAC case, it's important that the interrupt be completed earlier. But I think it is easy to come up with cases in which you don't want it completed earlier, e.g., a level interrupt on GPIO.

Since we probably don't want to go redesigning abstractions for this, I'd suggest just special casing this for now. That is, just special case on HMAC here and leave the logic for other interrupts unchanged.

phil-levis commented:
> 
> 
> It might be nice for someone else to take a look at the change to interrupt handling, as I know there can be some very subtle bugs there, but this looks good to me.

Added a comment. I think the issue is that one size does not fit all. While many interrupts don't care about the exact timing completion, some do. Given that generally interrupts should be able suppress subsequent interrupts in their handling (i.e., completed after the handler), the HMAC seems like a special case. Rather than come up with a change to the API that allows a handler to specify whether to complete, let's just special case HMAC for now?

alistair23 commented:
Fixed

alistair23 commented:
This morning in the OT call @phil-levis pointed out the same problem @hudson-ayers did above https://github.com/tock/tock/pull/2586#discussion_r634660361.

The issue is that to support the virtuliser all HMAC implementations need to implement HMAC256, HMAC384 and HMAC512. In OT's case there isn't support for the other 2 so we have an implementation that just returns an error. This goes against Rust's nice type system where we would get an error at build time if something isn't supported.

This is the same thing lots of console implementations do with receive.

Unfortunately there isn't really a good way to avoid this. The best other option (if I understood @phil-levis  correctly) is to have `hmac_256.rs`, `hmac_384.rs` and `hmac_512.rs` capsules. Then these all interact with a single `hmac_queue.rs` capsule that acts as mutex for the three implementations. This then has to sit on top of the same thing for a virtual digest to allow us to virtualise HMAC and SHA.

This would require a lot of code duplication and expose different HMAC capsules to user space increasing the complexity of the userspace implementation as well. I think although this current implementation isn't great in terms of compile time trait checks, it's much simpler and overall less bug prone then what I described above.

phil-levis commented:
I am not super-happy with the lack of type-checking, but it looks like only lowrisc supports this trait, I think we can punt on it for now. If it turns out we start having drivers that only support some subset of the widths we can revisit it; @alistair23 's point about additional code is very valid.

phil-levis commented:
bors r+
